webrtc/README.md

```diff
@@ -0,0 +1,417 @@
+
+It is an adaptation from the [QUIC RFC]. When in doubt on the semantics of
+these messages, consult the [QUIC RFC].
+
+``` proto
+syntax = "proto2";
+
+package webrtc.pb;
+
+message Message {
+  enum Flag {
+    // The sender will no longer send messages on the stream.
+    FIN = 0;
+    // The sender will no longer read messages on the stream. Incoming data is
+    // being discarded on receipt.
+    STOP_SENDING = 1;
+    // The sender abruptly terminates the sending part of the stream. The
+    // receiver can discard any data that it already received on that stream.
+    RESET_STREAM = 2;
+  }
+
+  optional Flag flag=1;
+
+  optional bytes message = 2;
+}
+```
+
+Note that in contrast to QUIC (see [QUIC RFC - 3.5 Solicited State
+Transitions](https://www.rfc-editor.org/rfc/rfc9000.html#section-3.5)) a libp2p
+WebRTC endpoint receiving a `STOP_SENDING` frame SHOULD NOT send a
+`RESET_STREAM` frame in reply. The `STOP_SENDING` frame is used for accurate
+accounting of the number of bytes sent for connection-level flow control in
+QUIC. The libp2p WebRTC message framing is not concerned with flow-control and
+thus does not need the `RESET_STREAM` frame to be send in reply to a
+`STOP_SENDING` frame.
+
+Encoded messages including their length prefix MUST NOT exceed 16kiB to support
+all major browsers. See ["Understanding message size
+limits"](https://developer.mozilla.org/en-US/docs/Web/API/WebRTC_API/Using_data_channels#understanding_message_size_limits).
+Implementations MAY choose to send smaller messages, e.g. to reduce delays
+sending _flagged_ messages.
+
+### Ordering
+
+Implementations MAY expose an unordered byte stream abstraction to the user by
+overriding the default value of `ordered` `true` to `false` when creating a new
+data channel via
+[`RTCPeerConnection.createDataChannel`](https://www.w3.org/TR/webrtc/#dom-peerconnection-createdatachannel).
+
+## Connection Security
+
+Note that the below uses the message framing described in
+[multiplexing](#multiplexing).
+
+While WebRTC offers confidentiality and integrity via TLS, one still needs to
+authenticate the remote peer by its libp2p identity.
+
+After [Connection Establishment](#connection-establishment):
+
+1. _A_ opens a WebRTC datachannel.
+
+2. _A_ starts a Noise `XX` handshake using _A_'s and _B_'s libp2p identity. See
+   [noise-libp2p](https://github.com/libp2p/specs/tree/master/noise).
+
+   Instead of exchanging the TLS certificate fingerprints on the established
+   Noise channel once the Noise handshake succeeded, _A_ and _B_ use the [Noise
+   Prologue](https://noiseprotocol.org/noise.html#prologue) mechanism, thus
+   saving one round trip.
+
+   More specifically _A_ and _B_ set the Noise _Prologue_ to
+   `libp2p-webrtc-noise:<FINGERPRINTS>` before starting the actual Noise
+   handshake. `<FINGERPRINTS>` is the concatenation of the two TLS
+   fingerprints of _A_ and _B_ in their multihash byte representation, sorted in
```

Comment:
What do we gain from sorting the fingerprints? It's not immediately obvious to me how to sort two fingerprint thats have unequal length, we'd have to specify this (you could either align on the first or the last byte, and then do a byte-wise comparison).

Given that we have clearly defined roles (one peer is the DTLS client, one is the DTLS server), why not just use `<client fingerprint><server fingerprint>`, thereby removing the ordering requirement.

Reply:
Note, if we decide on this, [the test vectors](
Conversation 259) need adjusting.

Reply:
Not sure if we'd need test vectors to show how to concat 3 strings though. On the other hand, it can't hurt.

Reply:
I don't have an opinion on whether to use a predefined order or whether to sort. Preference for the latter was raised in https://github.com/libp2p/specs/pull/412#discussion_r935347699. I am happy to switch it back to the former.

Reply:
> Not sure if we'd need test vectors to show how to concat 3 strings though. On the other hand, it can't hurt.

There are a few subtleties like different hash functions, the multihash representation, the sorting that might be easy to get wrong :)

Reply:
> I don't have an opinion on whether to use a predefined order or whether to sort. Preference for the latter was raised in [#412 (comment)](https://github.com/libp2p/specs/pull/412#discussion_r935347699). I am happy to switch it back to the former.

I don’t see any argument in https://github.com/libp2p/specs/pull/412#discussion_r935347699 for why sorting would be superior to just concatenating. All we need to achieve is that both peers agree on the order.
Given that it’s not necessary to sort, and we’d have to spend extra effort on defining how to sort unequal-length byte strings, concatenating seems like the strictly simpler solution.

Reply:
That sounds like an argument against sorting. Implementations really should take care to not mix up values between client and server. If sorting encourages implementations to act less carefully, we shouldn’t do it (we shouldn’t do in the first place because 1. it is underspecified and 2. it introduces unneeded complexity).

Reply:
I can see the argument for both sorting and not sorting. Based on the discussion above, I don't see how either is better than the other.

With that in mind, and given the fact that the js-libp2p, go-libp2p and rust-libp2p WebRTC implementations use sorting today (correct me if I am wrong), I recommend keeping this as is, thus sorting the fingerprints.

@marten-seemann in case you feel strongly about not-sorting, please comment here and I will change the specification.

Reply:
> Sorting the fingerprints makes this easier because you can pass them in either order and it will always be the same result.

You only need to implement correct ordering once. It's not that hard 😄 

> rust-libp2p WebRTC implementations use sorting today (correct me if I am wrong)

not the case for rust-libp2p

Reply:
> > rust-libp2p WebRTC implementations use sorting today (correct me if I am wrong)
> 
> not the case for rust-libp2p

@melekes am I misunderstanding the lines below?

``` rust
    let (a, b) = (a.to_bytes(), b.to_bytes());
    let (first, second) = if a < b { (a, b) } else { (b, a) };
    const PREFIX: &[u8] = b"libp2p-webrtc-noise:";
    let mut out = Vec::with_capacity(PREFIX.len() + first.len() + second.len());
```

https://github.com/libp2p/rust-libp2p/pull/2622/files#diff-038d467fe95855f8fed1e3122253ffb6bcb823fd74b83273c2ace20bf3fb3d4aR601-R604

Reply:
sorry, it was done by @thomaseizinger in https://github.com/melekes/rust-libp2p/pull/10

Reply:
> sorry, it was done by @thomaseizinger in [melekes/rust-libp2p#10](https://github.com/melekes/rust-libp2p/pull/10)

Reverted :blush:

Reply:
> > sorry, it was done by @thomaseizinger in [melekes/rust-libp2p#10](https://github.com/melekes/rust-libp2p/pull/10)
> 
> Reverted 😊

why?) i'm in favor of not sorting

Reply:
> > > sorry, it was done by @thomaseizinger in [melekes/rust-libp2p#10](https://github.com/melekes/rust-libp2p/pull/10)
> > 
> > 
> > Reverted blush
> 
> why?) i'm in favor of not sorting

Because it was a too eager commit, given that the current spec says sorting :)
Happy to change it once there is an agreement!

Reply:
> @marten-seemann in case you feel strongly about not-sorting, please comment here and I will change the specification.

I do. Sorry for that ;)
Thanks for making the change!

Reply:
730dca0 documents an order (_A_, then _B_) instead of sorting.

@ckousik, @John-LittleBearLabs, @melekes and @thomaseizinger please update the Golang, JS and Rust implementations accordingly.